src/compress.c

```diff
@@ -268,15 +387,198 @@ bool zseek_writer_close(zseek_writer_t *writer, char errbuf[ZSEEK_ERRBUF_SIZE])
     return !is_error;
 }
 
-bool zseek_write(zseek_writer_t *writer, const void *buf, size_t len,
+/**
+ * Flush, close and write current frame. This will block.
+ */
+static bool end_frame_lz4(zseek_writer_t *writer)
+{
+    // TODO: Communicate error info?
+
+    size_t ubuf_len = zseek_buffer_size(writer->buffer);
+    void *ubuf = zseek_buffer_data(writer->buffer);
+    assert(ubuf);
+
+    // Allocate output buffer
+#if LZ4_VERSION_NUMBER < 10800 // < v1.8.0
+    // NOTE: On lz4 < v1.8.0, acceleration is not exposed, so we fall back to
+    // using LZ4_compress_fast_extState directly.
+    const size_t BLOCK_SIZE_FIELD_SIZE = 4;
+    size_t cbuf_len = (size_t)LZ4_compressBound(writer->frame_uc); // data size
+    if (cbuf_len == 0) {
+        // fprintf(stderr, "input size too large");
+        goto fail;
+    }
+    cbuf_len += BLOCK_SIZE_FIELD_SIZE;  // block size field size
+    // NOTE: If we ever enable block checksum, account for it here (+4)
+    cbuf_len += LZ4F_compressBound(0, &writer->preferences);    // footer size
+#else
+    size_t cbuf_len = LZ4F_compressBound(writer->frame_uc,
+        &writer->preferences);
+#endif
+    void *cbuf = malloc(cbuf_len);
```

Comment:
There is no need to allocate an output buffer on each frame, you can resize the output buffer the same way you resize the input buffer.

Reply:
A very good point, thank you! Changing it.

Reply:
Actually, this is a more general issue: the resizable buffer being introduced here enables reusing the same buffer in other places as well (zstd too). This was more difficult previously because most of those buffers are not fixed in size.

In order to keep the scope and the volume of this changeset down, how about we change all of these in a separate PR?